Marlin/src/module/ft_motion.cpp

```diff
@@ -541,258 +433,261 @@ void FTMotion::setTrajectoryType(const TrajectoryType type) {
     case TrajectoryType::POLY5:       currentGenerator = &poly5Generator; break;
     case TrajectoryType::POLY6:       currentGenerator = &poly6Generator; break;
   }
-  currentGenerator->reset(); // Reset the selected generator
 }
 
 // Load / convert block data from planner to fixed-time control variables.
 // Called from FTMotion::loop() at the fetch of the next planner block.
-void FTMotion::loadBlockData(block_t * const current_block) {
-  // Cache the extruder index for this block
-  TERN_(DISTINCT_E_FACTORS, block_extruder_axis = E_AXIS_N(current_block->extruder));
+// Return whether a plan is available.
+bool FTMotion::plan_next_block() {
+  while (true) {
+
+    const bool had_block = !!stepper.current_block;
+    discard_planner_block_protected();                                  // Always clears stepper.current_block...
+    block_t * const current_block = planner.get_current_block();  // ...so get the current block from the queue
+
+    // The planner had a block and there was not another one?
+    const bool planner_finished = had_block && !current_block;
+    if (planner_finished) {
+      runoutBlock();
+      return true;
+    }
 
-  const float totalLength = current_block->millimeters,
-              oneOverLength = 1.0f / totalLength;
+    // There was never a block? Run out the plan and bail.
+    if (!current_block) {
+      currentGenerator->planRunout(0);
+      return false;
+    }
 
-  startPos = endPos_prevBlock;
-  const xyze_pos_t& moveDist = current_block->dist_mm;
-  ratio = moveDist * oneOverLength;
+    // Fetching this block for Stepper and for this loop
+    stepper.current_block = current_block;
 
-  const float mmps = totalLength / current_block->step_event_count; // (mm/step) Distance for each step
-  const float initial_speed = mmps * current_block->initial_rate;   // (mm/s) Start feedrate
-  const float final_speed = mmps * current_block->final_rate;       // (mm/s) End feedrate
+    // Handle sync blocks and skip others
+    if (current_block->is_sync_pos()) stepper._set_position(current_block->position);
+    if (current_block->is_sync()) continue;
```

Comment:
```suggestion
    // Handle sync blocks and skip others
    if (current_block->is_sync()) {     // Sync block?
      if (current_block->is_sync_pos()) // Position sync? Set the position.
        stepper._set_position(current_block->position);
      continue;
    }
```
It's mandatory to have` is_sync()` true when `is_sync_pos()` is true  but the reverse is not true. Old code could avoid one test.

Reply:
thanks, I'm mid refactoring and have added this now.